### PR TITLE
Delayed compilation of backwards pass to when backwards runs

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83367

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

File: functorch/functorch/_src/aot_autograd.py [hunk 1]
```diff
@@ -206,74 +206,62 @@ def new_fn(args):
...
+    joint_inputs = (flat_args, out)
+    with torch.set_grad_enabled(True):
+        fx_g = make_fx(joint_forward_backward, aot_config.decompositions)(*joint_inputs)
 
-                fx_g = make_fx(functionalize(fake_fn))(*joint_inputs)
+        if config.use_functionalize:
+            # Functionalize the foward backward graph. First create a
+            # fake fn to make functionalize happy
+            def fake_fn(primals, tangents):
+                return fx_g(primals, tangents)
```
Comment (zou3519) [inline functorch/functorch/_src/aot_autograd.py:231]: Why do we need a fake fn to make functionalize happy?

File: functorch/functorch/_src/aot_autograd.py [hunk 2]
```diff
@@ -206,74 +206,62 @@ def new_fn(args):
...
 
-        if config.debug_joint:
-            print(fx_g.code)
```
Comment (zou3519) [inline functorch/functorch/_src/aot_autograd.py:249]: I don't see a way to debug_joint anymore, is this intentional?

File: functorch/functorch/_src/aot_autograd.py [hunk 3]
```diff
@@ -282,8 +270,16 @@ def forward(ctx, *flat_tensor_args):
         def backward(ctx, *flat_args):
             contiguous_args = [t.contiguous() for t in flat_args]
             all_args = list(ctx.saved_tensors) + list(contiguous_args)
+            if CompiledFunction.compiled_bw is None:
+                with track_graph_compiling("backward", True):
+                    CompiledFunction.compiled_bw = aot_config.bw_compiler(
+                        bw_module, all_args
+                    )
```
Comment (zou3519) [inline functorch/functorch/_src/aot_autograd.py:277]: Add test? Shouldn't be too bad since we can pass separate compilers for forward and backward

**Review by zou3519 — APPROVED** [review]
let's add a test if we can, otherwise lgtm

File: functorch/functorch/_src/aot_autograd.py (hunk 1 above)
Comment (zou3519) [reply functorch/functorch/_src/aot_autograd.py:231]: (this smells like a bug in functionalize)

File: functorch/functorch/_src/aot_autograd.py (hunk 2 above)
Comment (Chillee) [reply functorch/functorch/_src/aot_autograd.py:249]: oops

File: functorch/functorch/_src/aot_autograd.py (hunk 1 above)
Comment (Chillee) [reply functorch/functorch/_src/aot_autograd.py:231]: not sure, but this was already there.

File: functorch/functorch/_src/aot_autograd.py (hunk 3 above)
Comment (Chillee) [reply functorch/functorch/_src/aot_autograd.py:278]: `test_grad_context` is already testing this (i.e. this change caused that test to fail haha).

Fixed that.

**Pull request closed without merging**